### PR TITLE
Copy 2023 files

### PR DESCRIPTION
**Pull request by fotnitfpro000**
added files for the root directory in the repo

### CODE REVIEW

File: Physics.py [hunk 1]
```diff
@@ -0,0 +1,189 @@
+from __future__ import annotations
```
Comment (auscompgeek) [inline Physics.py:1]: `physics.py` needs to be all lowercase

**Review by auscompgeek — CHANGES_REQUESTED** [review]

**Comment by james-ward** [comment]
Without going into the specifics of the PR, the way you have laid it out is perfect. You've got a commit for each change (file added), and it's just the single piece of work.
Well done.

**Comment by james-ward** [comment]
Is it worth copying over the `.github` directory so that we have some tests running? What do you think @auscompgeek @fotnitfpro000 ?

File: robot.py [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+
+        # stop rumble after time
+        if self.rumble_timer.hasElapsed(self.rumble_duration):
+            self.gamepad.setRumble(wpilib.XboxController.RumbleType.kBothRumble, 0)
+
+    def testInit(self) -> None:
+        pass
+        # disabled vision
+        # self.port_localizer.add_to_estimator = True
+        # self.starboard_localizer.add_to_estimator = True
```
Comment (mlists) [inline robot.py:110]: I think we should delete old commented code so we have a clean repo to start from

File: robot.py [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+
+    def testPeriodic(self) -> None:
+        # dpad_angle = self.gamepad.getPOV()
```
Comment (mlists) [inline robot.py:113]:
```suggestion
```

File: robot.py (hunk 2 above)
Comment (mlists) [inline robot.py:110]:
```suggestion
```

File: robot.py [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+
+        # Cancel any running controllers
+        if self.gamepad.getBackButtonPressed():
+            self.cancel_controllers()
+
+        # Tick the controllers
+        # These will only do anything if engage() has been called on them
```
Comment (mlists) [inline robot.py:120]: We are not actually doing this at the moment
```suggestion
```

File: robot.py [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+        self.movement.done()
+
+    def disabledPeriodic(self) -> None:
+        self.chassis.update_odometry()
+
+    def autonomousInit(self) -> None:
+        pass
+        # disabled vision
+        # self.port_localizer.add_to_estimator = True
+        # self.starboard_localizer.add_to_estimator = True
```
Comment (mlists) [inline robot.py:134]:
```suggestion
```

**Review by mlists — CHANGES_REQUESTED** [review]

File: robot.py [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+            0.95, self.event_loop
+        ).rising()
+        self.left_trigger_down_half = self.gamepad.leftTrigger(
+            0.05, self.event_loop
+        ).rising()
+        self.left_trigger_up = self.gamepad.leftTrigger(0.95, self.event_loop).falling()
+
+        self.rumble_timer = wpilib.Timer()
+        self.rumble_timer.start()
+        self.rumble_duration = 0.0
```
[the remark this answers is not in the document]
Comment (auscompgeek) [reply robot.py:49]: Let's do a reorg of the code later.

File: controllers/movement.py [hunk 7]
```diff
@@ -0,0 +1,248 @@
...
+        vx, vy: velocities in m/s
+        vz: rotational velocity in rad/s
+        local: drive reletive to robot frame rather than global frame
+        override: override driver lockout
+        """
+        if not self.inputs_lock or override:
+            # if is_red():
+            #     vx = -vx
+            #     vy = -vy
+
```
Comment (auscompgeek) [inline controllers/movement.py:243]:
```suggestion
```

File: controllers/movement.py [hunk 8]
```diff
@@ -0,0 +1,248 @@
...
+    HolonomicDriveController,
+    PIDController,
+    ProfiledPIDControllerRadians,
+)
+from wpimath.spline import Spline3
+import math
+from wpilib import Field2d
+
+# from utilities.game import is_red
+
```
Comment (auscompgeek) [inline controllers/movement.py:31]:
```suggestion
```

File: robot.py [hunk 9]
```diff
@@ -0,0 +1,106 @@
...
+        ).falling()
+
+        # Left trigger events
+        self.left_trigger_down_full = self.gamepad.leftTrigger(
+            0.95, self.event_loop
+        ).rising()
+        self.left_trigger_down_half = self.gamepad.leftTrigger(
+            0.05, self.event_loop
+        ).rising()
+        self.left_trigger_up = self.gamepad.leftTrigger(0.95, self.event_loop).falling()
```
Comment (auscompgeek) [inline robot.py:44]: nit: These seem completely unused. Should we delete?

**Review by auscompgeek — COMMENTED** [review]
more dead code

File: robot.py (hunk 9 above)
Comment (outsidermm) [reply robot.py:44]: I think we can delete the trigger events if we are just trying to make the drive base run on joysticks.

Comment (auscompgeek) [reply robot.py:44]: Good call. We don't really know what sort of control scheme (or HID) the next drive team will prefer, given we haven't finalised the drive team yet.

Comment (auscompgeek) [reply robot.py:44]:
```suggestion
```

**Review by auscompgeek — APPROVED** [review]
LGTM once the tests and CI are added.

**Comment by outsidermm** [comment]
> LGTM once the tests and CI are added.

before we pull don't we have to make sure the drive base runs in sim

**Comment by auscompgeek** [comment]
Oh, good point. I ran the tests/sim but didn't actually test that it moves.

Just tried it and it isn't moving.

File: robot.py [hunk 10]
```diff
@@ -0,0 +1,86 @@
...
+from controllers.movement import Movement
+
+from utilities.scalers import rescale_js
+
+
+class MyRobot(magicbot.MagicRobot):
+    # Components
+    chassis: Chassis
+
+    movement: Movement
```
Comment (auscompgeek) [inline robot.py:17]: Controllers need to run before the components they control.
```suggestion
    # Controllers
    movement: Movement

    # Components
    chassis: Chassis
```

File: robot.py [hunk 11]
```diff
@@ -0,0 +1,86 @@
...
+        self.rumble_timer = wpilib.Timer()
+        self.rumble_timer.start()
+        self.rumble_duration = 0.0
+
+        self.field = wpilib.Field2d()
+        wpilib.SmartDashboard.putData(self.field)
+        self.target_node = self.field.getObject("target_node")
+
+        self.last_dpad = -1
+
```
Comment (auscompgeek) [inline robot.py:35]: also unused
```suggestion
```

File: robot.py [hunk 12]
```diff
@@ -0,0 +1,86 @@
...
+        self.rumble_for(0.4, 0.1)
+
+    def long_rumble(self):
+        self.rumble_for(0.8, 0.3)
+
+    def teleopInit(self) -> None:
+        pass
+
+    def teleopPeriodic(self) -> None:
+        self.event_loop.poll()
```
Comment (auscompgeek) [inline robot.py:51]: can't poll an event loop we removed, oops
```suggestion
```

**Comment by james-ward** [comment]
Running the manual control through a state machine is an abomination and I think it should go too.

**Comment by outsidermm** [comment]
> Running the manual control through a state machine is an abomination and I think it should go too.

movement.py is removed, and so is the state machine; the basic drive base teleop control is now in robot.py if that's what you mean.

**Comment by james-ward** [comment]
Looks good to me. I can't confirm it moves in the sim where I am now, but if it does, LGTM.

**Comment by outsidermm** [comment]
> Looks good to me. I can't confirm it moves in the sim where I am now, but if it does, LGTM.

should I try to fix the CI / mypy, I'm not sure what exactly is causing that. The robot does move in my sim.

**Comment by mlists** [comment]
I think you are missing your __init__.pys in each folder which is causing that CI error

**Comment by outsidermm** [comment]
can you @mlists , please review and see if it looks good to merge into the main? thanks!
by adding the __init__.py did fix the mypy issue

**Review by mlists — APPROVED** [review]
Looks good to me

**Pull request merged**